File: LeetCode/[E] - 21. Merge Two Sorted Lists/solution.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        if list1 is None and list2 is None:
            return None
        head = ListNode()
        list3 = head
        while list1 and list2:
            if list1.val <= list2.val:
                list3.next = list1
                list1 = list1.next
            else:
                list3.next = list2
                list2 = list2.next
            list3 = list3.next
        while list1:
            list3.next = list1
            list3 = list3.next
            list1 = list1.next
        while list2:
            list3.next = list2
            list3 = list3.next
            list2 = list2.next
        return head.next

File: LeetCode/[E] - 21. Merge Two Sorted Lists/solution.go
package main

// ListNode Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil && list2 == nil {
		return nil
	}
	head := ListNode{}
	list3 := &head
	for list1 != nil && list2 != nil {
		if list1.Val <= list2.Val {
			list3.Next = list1
			list1 = list1.Next
		} else {
			list3.Next = list2
			list2 = list2.Next
		}
		list3 = list3.Next
	}
	for list1 != nil {
		list3.Next = list1
		list3 = list3.Next
		list1 = list1.Next
	}
	for list2 != nil {
		list3.Next = list2
		list3 = list3.Next
		list2 = list2.Next
	}
	return head.Next
}
